architecture-guide.md

```diff
@@ -0,0 +1,425 @@
+
+  Overly parameterized components are fragile and not easy to maintain. Big set
+  of switches, altering the appearance and behavior of the component, introduces
+  a lot of conditional statements in the internal implementation. Finally,
+  it's difficult to assure that all combinations are valid or specify which ones
+  are valid and desired.
+
+- Action signals (e.g. click on delegate) should not provide metadata from the
+  model. Only `index` or unique `key` should be an argument. The necessary
+  metadata should be fetched on the signal receiver side.
+
+  ```qml
+  signal collectibleClicked(int chainId, string contractAddress, 
+                            string tokenId, string uid, int tokenType) // BAD
+  signal collectibleClicked(var collectible)                           // BAD
+
+  signal collectibleClicked(string key)                                // OK
+  ```
+  
+  Otherwise, the component unnecessarily bypasses some values from the model.
+  Callers may also want to access roles which are not used for displaying.
+  It leads to creating unnecessary requirements for the input model.
+
+- use consistently `key` as a unique identifier of the model, even if there are
+  other roles with unique values.
+  
+  On the UI side, `key` should be used only for identification, with no other
+  assumptions and usage. Thanks to that, content of that role can be freely
+  changed on the backend with no implications to the UI side. Even if `address`
+  is unique, a separate `key` role should be used (providing the same content as
+  `address` role).
+  
+  Other good name for that role - `id` - doesn't fit well in qml environment as
+  it's reserved keyword. Defining such role with `ListElement` in tests or
+  Storybook would be not possible.
+  
+- `model.rowCount()` should not be used in bindings.
+  
+  This call provides count only once when it's called, and the expression won't
+  be re-evaluated when count is changed. Solution is to use the attached property
+  `ModelCount` (`model.ModelCount.count`) context property instead. However,
+  in signal handling, using `rowCount()` is fully correct.
+
+- `model.count` should not be used on models taken from outside as a dependency.
+  
+  `count` property is not a part of the `QAbstractItemModel` interface. Proxy
+  models and backend models may not have that property defined.
+
+- Objects holding the whole model's row should not be used in bindings.
+
+  ```qml
+  readonly property var selectedAccount: ModelUtils.get(store.accounts, 0) // BAD
+  ```
+
+  Such objects may be deleted at any time, e.g. because of model reset. The
+  expression will not be automatically re-evaluated to take the current first item.
+
+- Avoid `QML`'s dynamic scoping. Refer for details
+  [here](https://doc.qt.io/qt-5/qtqml-documents-scope.html#component-instance-hierarchy).
+
+- Follow [C++ Core Guidelines](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines)
+  in C++ code.
+
+## Stores
```

Comment:
As we are talking here about `Stores` maybe it makes sense to add an ideal example of the structure we should have or we want to have in future:

- Only 1 `RootStore` that will instantiate the rest of "context" stores
- Example for **Wallet** context:
  - **WalletStore** (this should be the "root wallet" store): This should instantiate the rest of wallet related stores and maybe have some global properties (nothing else).
    - _TransactionsStore_: Just specific for interacting with transactions.
    - _AssetsStore_: Just specific for interacting with assets.
    - _CollectiblesStore_: Just specific for interacting with collectibles.
    ...    
- _RootStore_ instantiated in _appMain_ and following the statements of "passing the stores as dependences and not using sigletons", pass the needed stores from _appMain_ to the deeper UI components.

Reply:
Good point, I added some additional note about that but in general the structure of the stores is a subject of further investigation and agreements.